### PR TITLE
Add load file by regex and wildcard

### PR DESCRIPTION
**Pull request by dankwart-de**

### CODE REVIEW

**Comment by dankwart-de** [comment]
Review after #274

File: RoddyCore/src/de/dkfz/roddy/execution/io/fs/ShellCommandSet.groovy [hunk 1]
```diff
@@ -69,9 +70,40 @@ abstract class ShellCommandSet {
...
+        this.getListFullDirectoriesContentRecursivelyCommand(directories, depth, onlyDirectories ? FileType.DIRECTORIES : FileType.ANY, true)
+    }
+
+    /**
+     * API Level 3.2+
+     * Return a command to create a tree of files and / or directories for a range of directories.
+     * @param directories   The base folders to look into
+     * @param depth         The maximum depth to go into for each directory
+     *                      Be careful: If depth is not set properly, -1 (no depth) will be used for every directory!
+     * @param selectedType  ANY, FILES or DIRECTORIES
```
[the remark this answers is not in the document]
Comment (dankwart-de) [reply RoddyCore/src/de/dkfz/roddy/execution/io/fs/ShellCommandSet.groovy]: In that case I reuse an existing Java / Groovy class. Can't change it, as it is not mine.

File: RoddyCore/src/de/dkfz/roddy/execution/io/fs/FileSystemAccessProvider.groovy [hunk 2]
```diff
@@ -293,6 +294,39 @@ public class FileSystemAccessProvider {
...
+        } as List<File>
+    }
+
+    /**
+     * Scope for regular expressions. Used in listFilesUsingRegex
+     */
+    enum RegexScope {
+        FullPath,
+        RelativePath,
+        Filename
```
[the remark this answers is not in the document]
Comment (dankwart-de) [reply RoddyCore/src/de/dkfz/roddy/execution/io/fs/FileSystemAccessProvider.groovy]: It is the files base name as returned by $(basename file...). But the File.name() comes from the Java world. So I'd stick to Filename.

File: RoddyCore/src/de/dkfz/roddy/execution/io/fs/FileSystemAccessProvider.groovy [hunk 3]
```diff
@@ -293,6 +294,39 @@ public class FileSystemAccessProvider {
...
+        result.resultLines.collect {
+            new File(it)
+        } as List<File>
+    }
+
+    /**
+     * Scope for regular expressions. Used in listFilesUsingRegex
+     */
+    enum RegexScope {
+        FullPath,
```
[the remark this answers is not in the document]
Comment (dankwart-de) [reply RoddyCore/src/de/dkfz/roddy/execution/io/fs/FileSystemAccessProvider.groovy]: FullPath => AbsolutePath 
RelativePath => Path relative to the folder in which is searched

File: RoddyCore/test/src/de/dkfz/roddy/execution/io/fs/BashCommandSetSpecification.groovy [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package de.dkfz.roddy.execution.io.fs
```
[the remark this answers is not in the document]
Comment (dankwart-de) [reply RoddyCore/test/src/de/dkfz/roddy/execution/io/fs/BashCommandSetSpecification.groovy]: Took its time :D

File: RoddyCore/src/de/dkfz/roddy/core/Workflow.groovy [hunk 5]
```diff
@@ -281,6 +282,20 @@ abstract class Workflow {
...
+     * API Level 3.2+
+     */
+    final List<BaseFile> getSourceFilesUsingWildcards(File path, String wildcards, String _class = BaseFile.STANDARD_FILE_CLASS) {
+        BaseFile.getSourceFilesUsingWildcards(context, path, wildcards, _class)
+    }
+
+    /**
+     * API Level 3.2+
+     */
+    final List<BaseFile> getSourceFilesUsingRegex(File path, String regex, FileSystemAccessProvider.RegexScope scope, String _class = BaseFile.STANDARD_FILE_CLASS) {
```
[the remark this answers is not in the document]
Comment (dankwart-de) [reply RoddyCore/src/de/dkfz/roddy/core/Workflow.groovy]: What I found is new Pattern(/somePattern/, aNumber). For me this does not look to nice. Should we encapsulate it in a custom class? 
For Wildcard. There are e.g. WildcardFilenameFilters. But I don't think they are the right choice for our use-case.

[the remark this answers is not in the document]
Comment (vinjana) [reply RoddyCore/src/de/dkfz/roddy/core/Workflow.groovy]: Hm. Not sure. We had something similar for identifiers. To make sure the right method is chosen automatically, I created a number of Identifier subclasses. Also here we could just make wrappers around String. I am not sure whether such an approach would be kind of overkill here. Somehow this method is part of the outward interface, right? We should certainly be careful here.

Can you explain, what is the problem with using the Regex class?

File: RoddyCore/src/de/dkfz/roddy/execution/io/fs/FileSystemAccessProvider.groovy [hunk 6]
```diff
@@ -293,6 +294,39 @@ public class FileSystemAccessProvider {
...
+        ExecutionResult result = ExecutionService.instance.execute(commandSet.getFindFilesUsingWildcardsCommand(baseFolder, wildcards))
+        result.resultLines.collect {
+            new File(it)
+        } as List<File>
+    }
+
+    /**
+     * Scope for regular expressions. Used in listFilesUsingRegex
+     */
+    enum RegexScope {
```
[the remark this answers is not in the document]
Comment (vinjana) [reply RoddyCore/src/de/dkfz/roddy/execution/io/fs/FileSystemAccessProvider.groovy]: or PathSearchScope. Something that expresses better, what this enum is supposed to be applied to.

File: RoddyCore/src/de/dkfz/roddy/execution/io/fs/FileSystemAccessProvider.groovy (hunk 2 above)
[the remark this answers is not in the document]
Comment (vinjana) [reply RoddyCore/src/de/dkfz/roddy/execution/io/fs/FileSystemAccessProvider.groovy]: Then change `FullPath` to `AbsolutePath` -- in the Java world this is `file.absolutePath`.

File: RoddyCore/src/de/dkfz/roddy/execution/io/fs/ShellCommandSet.groovy [hunk 7]
```diff
@@ -69,9 +70,40 @@ abstract class ShellCommandSet {
...
+        this.getListFullDirectoryContentRecursivelyCommand(f, depth, onlyDirectories ? FileType.DIRECTORIES : FileType.ANY, true)
+    }
+
+    /**
+     * API Level 3.2+
+     * Return a command to create a tree of files and / or directories for a given directory
+     * @param directory     The base folder to look into
+     * @param depth         The maximum depth to go into
+     * @param selectedType  ANY, FILES or DIRECTORIES
+     * @param complexList   Get more info than the filename (e.g. size, owner etc.). Highly dependent on the backend!
```
[the remark this answers is not in the document]
Comment (vinjana) [reply RoddyCore/src/de/dkfz/roddy/execution/io/fs/ShellCommandSet.groovy]: Right, find in Mac is different. Should we then not move this method to the BashCommandSet? (even though it is not Bash, but Linux Core Utils that can also be used from other shells?)

Also ..., we don't have any other subclass of ShellCommandSet but BashCommandSet.

File: RoddyCore/src/de/dkfz/roddy/execution/io/fs/ShellCommandSet.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (vinjana) [reply RoddyCore/src/de/dkfz/roddy/execution/io/fs/ShellCommandSet.groovy]: o.k.

File: RoddyCore/src/de/dkfz/roddy/execution/io/fs/ShellCommandSet.groovy (hunk 7 above)
[the remark this answers is not in the document]
Comment (dankwart-de) [reply RoddyCore/src/de/dkfz/roddy/execution/io/fs/ShellCommandSet.groovy]: Well here it is only the base class. I adapted the comments so it gets a lot clearer.

**Comment by dankwart-de** [comment]
Discussed with vinjana, will merge

**Pull request merged**